File: datastructures-go/hash_table.go
package datastructures

import (
	"encoding/binary"
	"fmt"
	"io"
)

type SlotState uint8

const (
	Empty SlotState = iota
	Filled
	Deleted
)

const maxLoadFactor = 0.6

type HashTable struct {
	keys    []int
	values  []int
	states  []SlotState
	capacity int
	count    int
}

func NewHashTable(capacity ...int) *HashTable {
	cap := 8
	if len(capacity) > 0 && capacity[0] > 0 {
		cap = capacity[0]
	}
	
	ht := &HashTable{
		capacity: nextPow2(cap),
	}
	ht.keys = make([]int, ht.capacity)
	ht.values = make([]int, ht.capacity)
	ht.states = make([]SlotState, ht.capacity)
	
	for i := range ht.states {
		ht.states[i] = Empty
	}
	
	return ht
}

func nextPow2(n int) int {
	p := 1
	for p < n {
		p <<= 1
	}
	return p
}

func (ht *HashTable) hash(key int) int {
	x := uint32(key)
	x *= 2654435761
	return int(x & uint32(ht.capacity-1))
}

func (ht *HashTable) rehash(newCapacity int) {
	newCapacity = nextPow2(newCapacity)
	if newCapacity < 2 {
		newCapacity = 2
	}
	
	oldKeys := ht.keys
	oldValues := ht.values
	oldStates := ht.states
	oldCap := ht.capacity
	
	ht.keys = make([]int, newCapacity)
	ht.values = make([]int, newCapacity)
	ht.states = make([]SlotState, newCapacity)
	ht.capacity = newCapacity
	ht.count = 0
	
	for i := range ht.states {
		ht.states[i] = Empty
	}
	
	for i := 0; i < oldCap; i++ {
		if oldStates[i] == Filled {
			ht.Insert(oldKeys[i], oldValues[i])
		}
	}
}

func (ht *HashTable) Insert(key, value int) {
	if float64(ht.count+1)/float64(ht.capacity) > maxLoadFactor {
		ht.rehash(ht.capacity * 2)
	}
	
	idx := ht.hash(key)
	firstDeleted := -1
	
	for probe := 0; probe < ht.capacity; probe++ {
		pos := (idx + probe) & (ht.capacity - 1)
		
		if ht.states[pos] == Filled && ht.keys[pos] == key {
			ht.values[pos] = value
			return
		}
		
		if ht.states[pos] == Deleted && firstDeleted == -1 {
			firstDeleted = pos
		}
		
		if ht.states[pos] == Empty {
			target := pos
			if firstDeleted != -1 {
				target = firstDeleted
			}
			ht.keys[target] = key
			ht.values[target] = value
			ht.states[target] = Filled
			ht.count++
			return
		}
	}
	
	ht.rehash(ht.capacity * 2)
	ht.Insert(key, value)
}

func (ht *HashTable) Contains(key int) bool {
	return ht.Get(key) != -1
}

func (ht *HashTable) Get(key int) int {
	idx := ht.hash(key)
	
	for probe := 0; probe < ht.capacity; probe++ {
		pos := (idx + probe) & (ht.capacity - 1)
		
		if ht.states[pos] == Empty {
			return -1
		}
		
		if ht.states[pos] == Filled && ht.keys[pos] == key {
			return ht.values[pos]
		}
	}
	
	return -1
}

func (ht *HashTable) Remove(key int) {
	idx := ht.hash(key)
	
	for probe := 0; probe < ht.capacity; probe++ {
		pos := (idx + probe) & (ht.capacity - 1)
		
		if ht.states[pos] == Empty {
			return
		}
		
		if ht.states[pos] == Filled && ht.keys[pos] == key {
			ht.states[pos] = Deleted
			ht.count--
			return
		}
	}
}

func (ht *HashTable) Size() int {
	return ht.count
}

func (ht *HashTable) Items() []struct{ Key, Value int } {
	items := make([]struct{ Key, Value int }, 0, ht.count)
	
	for i := 0; i < ht.capacity; i++ {
		if ht.states[i] == Filled {
			items = append(items, struct{ Key, Value int }{
				Key:   ht.keys[i],
				Value: ht.values[i],
			})
		}
	}
	
	return items
}

func (ht *HashTable) SerializeBinary(w io.Writer) error {
	items := ht.Items()
	length := uint32(len(items))
	
	if err := binary.Write(w, binary.LittleEndian, length); err != nil {
		return err
	}
	
	for _, item := range items {
		if err := binary.Write(w, binary.LittleEndian, int32(item.Key)); err != nil {
			return err
		}
		if err := binary.Write(w, binary.LittleEndian, int32(item.Value)); err != nil {
			return err
		}
	}
	
	return nil
}

func (ht *HashTable) DeserializeBinary(r io.Reader) error {
	var length uint32
	if err := binary.Read(r, binary.LittleEndian, &length); err != nil {
		return err
	}
	
	ht.rehash(nextPow2(int(length)*2 + 8))
	
	for i := uint32(0); i < length; i++ {
		var key, value int32
		if err := binary.Read(r, binary.LittleEndian, &key); err != nil {
			return err
		}
		if err := binary.Read(r, binary.LittleEndian, &value); err != nil {
			return err
		}
		
		ht.Insert(int(key), int(value))
	}
	
	return nil
}

func (ht *HashTable) SerializeText(w io.Writer) error {
	items := ht.Items()
	
	if _, err := fmt.Fprintf(w, "%d\n", len(items)); err != nil {
		return err
	}
	
	for _, item := range items {
		if _, err := fmt.Fprintf(w, "%d %d\n", item.Key, item.Value); err != nil {
			return err
		}
	}
	
	return nil
}

func (ht *HashTable) DeserializeText(r io.Reader) error {
	var length int
	if _, err := fmt.Fscanf(r, "%d\n", &length); err != nil {
		return err
	}
	
	ht.rehash(nextPow2(length*2 + 8))
	
	for i := 0; i < length; i++ {
		var key, value int
		if _, err := fmt.Fscanf(r, "%d %d\n", &key, &value); err != nil {
			return err
		}
		
		ht.Insert(key, value)
	}
	
	return nil
}

type CuckooHashTable struct {
	keys1   []int
	vals1   []int
	used1   []bool
	
	keys2   []int
	vals2   []int
	used2   []bool
	
	capacity int
	count    int
}

func NewCuckooHashTable(capacity ...int) *CuckooHashTable {
	cap := 8
	if len(capacity) > 0 && capacity[0] > 0 {
		cap = capacity[0]
	}
	
	capacityInt := 1
	for capacityInt < cap {
		capacityInt <<= 1
	}
	if capacityInt < 2 {
		capacityInt = 2
	}
	
	cht := &CuckooHashTable{
		capacity: capacityInt,
	}
	
	cht.keys1 = make([]int, cht.capacity)
	cht.vals1 = make([]int, cht.capacity)
	cht.used1 = make([]bool, cht.capacity)
	
	cht.keys2 = make([]int, cht.capacity)
	cht.vals2 = make([]int, cht.capacity)
	cht.used2 = make([]bool, cht.capacity)
	
	return cht
}

func (cht *CuckooHashTable) hash1(key int) int {
	x := uint32(key)
	x *= 2654435761
	return int(x & uint32(cht.capacity-1))
}

func (cht *CuckooHashTable) hash2(key int) int {
	x := uint32(key)
	x ^= x >> 16
	x *= 2246822519
	x ^= x >> 13
	return int(x & uint32(cht.capacity-1))
}

func (cht *CuckooHashTable) rehash(newCapacity int) {
	current := cht.Items()
	
	cht.capacity = 1
	for cht.capacity < newCapacity {
		cht.capacity <<= 1
	}
	if cht.capacity < 2 {
		cht.capacity = 2
	}
	
	cht.keys1 = make([]int, cht.capacity)
	cht.vals1 = make([]int, cht.capacity)
	cht.used1 = make([]bool, cht.capacity)
	
	cht.keys2 = make([]int, cht.capacity)
	cht.vals2 = make([]int, cht.capacity)
	cht.used2 = make([]bool, cht.capacity)
	
	cht.count = 0
	
	for _, item := range current {
		cht.Insert(item.Key, item.Value)
	}
}

func (cht *CuckooHashTable) Contains(key int) bool {
	return cht.Get(key) != -1
}

func (cht *CuckooHashTable) Get(key int) int {
	i1 := cht.hash1(key)
	if cht.used1[i1] && cht.keys1[i1] == key {
		return cht.vals1[i1]
	}
	
	i2 := cht.hash2(key)
	if cht.used2[i2] && cht.keys2[i2] == key {
		return cht.vals2[i2]
	}
	
	return -1
}

func (cht *CuckooHashTable) Insert(key, value int) {
	if cht.Contains(key) {
		i1 := cht.hash1(key)
		if cht.used1[i1] && cht.keys1[i1] == key {
			cht.vals1[i1] = value
			return
		}
		
		i2 := cht.hash2(key)
		if cht.used2[i2] && cht.keys2[i2] == key {
			cht.vals2[i2] = value
			return
		}
	}
	
	if float64(cht.count+1)/float64(cht.capacity) > 0.45 {
		cht.rehash(cht.capacity * 2)
	}
	
	curKey := key
	curVal := value
	
	const maxKicks = 32
	for kick := 0; kick < maxKicks; kick++ {
		i1 := cht.hash1(curKey)
		if !cht.used1[i1] {
			cht.used1[i1] = true
			cht.keys1[i1] = curKey
			cht.vals1[i1] = curVal
			cht.count++
			return
		}
		
		curKey, cht.keys1[i1] = cht.keys1[i1], curKey
		curVal, cht.vals1[i1] = cht.vals1[i1], curVal
		
		i2 := cht.hash2(curKey)
		if !cht.used2[i2] {
			cht.used2[i2] = true
			cht.keys2[i2] = curKey
			cht.vals2[i2] = curVal
			cht.count++
			return
		}
		
		curKey, cht.keys2[i2] = cht.keys2[i2], curKey
		curVal, cht.vals2[i2] = cht.vals2[i2], curVal
	}
	
	cht.rehash(cht.capacity * 2)
	cht.Insert(curKey, curVal)
}

func (cht *CuckooHashTable) Remove(key int) {
	i1 := cht.hash1(key)
	if cht.used1[i1] && cht.keys1[i1] == key {
		cht.used1[i1] = false
		cht.count--
		return
	}
	
	i2 := cht.hash2(key)
	if cht.used2[i2] && cht.keys2[i2] == key {
		cht.used2[i2] = false
		cht.count--
		return
	}
}

func (cht *CuckooHashTable) Size() int {
	return cht.count
}

func (cht *CuckooHashTable) Items() []struct{ Key, Value int } {
	items := make([]struct{ Key, Value int }, 0, cht.count)
	
	for i := 0; i < cht.capacity; i++ {
		if cht.used1[i] {
			items = append(items, struct{ Key, Value int }{
				Key:   cht.keys1[i],
				Value: cht.vals1[i],
			})
		}
	}
	
	for i := 0; i < cht.capacity; i++ {
		if cht.used2[i] {
			items = append(items, struct{ Key, Value int }{
				Key:   cht.keys2[i],
				Value: cht.vals2[i],
			})
		}
	}
	
	return items
}

func (cht *CuckooHashTable) SerializeBinary(w io.Writer) error {
	items := cht.Items()
	length := uint32(len(items))
	
	if err := binary.Write(w, binary.LittleEndian, length); err != nil {
		return err
	}
	
	for _, item := range items {
		if err := binary.Write(w, binary.LittleEndian, int32(item.Key)); err != nil {
			return err
		}
		if err := binary.Write(w, binary.LittleEndian, int32(item.Value)); err != nil {
			return err
		}
	}
	
	return nil
}

func (cht *CuckooHashTable) DeserializeBinary(r io.Reader) error {
	var length uint32
	if err := binary.Read(r, binary.LittleEndian, &length); err != nil {
		return err
	}
	
	cht.rehash(int(length)*4 + 8)
	
	for i := uint32(0); i < length; i++ {
		var key, value int32
		if err := binary.Read(r, binary.LittleEndian, &key); err != nil {
			return err
		}
		if err := binary.Read(r, binary.LittleEndian, &value); err != nil {
			return err
		}
		
		cht.Insert(int(key), int(value))
	}
	
	return nil
}

func (cht *CuckooHashTable) SerializeText(w io.Writer) error {
	items := cht.Items()
	
	if _, err := fmt.Fprintf(w, "%d\n", len(items)); err != nil {
		return err
	}
	
	for _, item := range items {
		if _, err := fmt.Fprintf(w, "%d %d\n", item.Key, item.Value); err != nil {
			return err
		}
	}
	
	return nil
}

func (cht *CuckooHashTable) DeserializeText(r io.Reader) error {
	var length int
	if _, err := fmt.Fscanf(r, "%d\n", &length); err != nil {
		return err
	}
	
	cht.rehash(length*4 + 8)
	
	for i := 0; i < length; i++ {
		var key, value int
		if _, err := fmt.Fscanf(r, "%d %d\n", &key, &value); err != nil {
			return err
		}
		
		cht.Insert(key, value)
	}
	
	return nil
}

File: C++/hash_table.cpp
#include "hash_table.h"

#include <algorithm>
#include <cstdint>
#include <istream>
#include <ostream>
#include <stdexcept>
#include <utility>
#include <vector>

namespace {
constexpr double kMaxLoad = 0.6;

int clamp_min_pow2(int n) {
    if (n < 2) return 2;
    return n;
}
}


HashTable::HashTable(int capacity) {
    capacity_ = next_pow2(clamp_min_pow2(capacity));
    keys_ = new int[capacity_];
    values_ = new int[capacity_];
    states_ = new SlotState[capacity_];
    std::fill(states_, states_ + capacity_, SlotState::Empty);
}

HashTable::~HashTable() {
    delete[] keys_;
    delete[] values_;
    delete[] states_;
}

int HashTable::next_pow2(int n) const {
    int p = 1;
    while (p < n) p <<= 1;
    return p;
}

int HashTable::hash(int key) const {
    std::uint32_t x = static_cast<std::uint32_t>(key);
    x *= 2654435761u;
    return static_cast<int>(x & static_cast<std::uint32_t>(capacity_ - 1));
}

void HashTable::rehash(int new_capacity) {
    new_capacity = next_pow2(clamp_min_pow2(new_capacity));
    int* old_keys = keys_;
    int* old_vals = values_;
    SlotState* old_states = states_;
    int old_cap = capacity_;

    keys_ = new int[new_capacity];
    values_ = new int[new_capacity];
    states_ = new SlotState[new_capacity];
    capacity_ = new_capacity;
    count_ = 0;
    std::fill(states_, states_ + capacity_, SlotState::Empty);

    for (int i = 0; i < old_cap; ++i) {
        if (old_states[i] == SlotState::Filled) {
            insert(old_keys[i], old_vals[i]);
        }
    }

    delete[] old_keys;
    delete[] old_vals;
    delete[] old_states;
}

void HashTable::insert(int key, int value) {
    if (static_cast<double>(count_ + 1) / static_cast<double>(capacity_) > kMaxLoad) {
        rehash(capacity_ * 2);
    }

    int idx = hash(key);
    int first_deleted = -1;

    for (int probe = 0; probe < capacity_; ++probe) {
        const int pos = (idx + probe) & (capacity_ - 1);

        if (states_[pos] == SlotState::Filled && keys_[pos] == key) {
            values_[pos] = value;
            return;
        }
        if (states_[pos] == SlotState::Deleted && first_deleted == -1) {
            first_deleted = pos;
        }
        if (states_[pos] == SlotState::Empty) {
            const int target = (first_deleted != -1) ? first_deleted : pos;
            keys_[target] = key;
            values_[target] = value;
            states_[target] = SlotState::Filled;
            ++count_;
            return;
        }
    }

    rehash(capacity_ * 2);
    insert(key, value);
}

bool HashTable::contains(int key) const { return get(key) != -1; }

int HashTable::get(int key) const {
    int idx = hash(key);
    for (int probe = 0; probe < capacity_; ++probe) {
        const int pos = (idx + probe) & (capacity_ - 1);
        if (states_[pos] == SlotState::Empty) {
            return -1;
        }
        if (states_[pos] == SlotState::Filled && keys_[pos] == key) {
            return values_[pos];
        }
    }
    return -1;
}

void HashTable::remove(int key) {
    int idx = hash(key);
    for (int probe = 0; probe < capacity_; ++probe) {
        const int pos = (idx + probe) & (capacity_ - 1);
        if (states_[pos] == SlotState::Empty) {
            return;
        }
        if (states_[pos] == SlotState::Filled && keys_[pos] == key) {
            states_[pos] = SlotState::Deleted;
            --count_;
            return;
        }
    }
}

std::vector<std::pair<int, int>> HashTable::items() const {
    std::vector<std::pair<int, int>> out;
    out.reserve(static_cast<std::size_t>(count_));
    for (int i = 0; i < capacity_; ++i) {
        if (states_[i] == SlotState::Filled) {
            out.emplace_back(keys_[i], values_[i]);
        }
    }
    return out;
}

void HashTable::serialize_binary(std::ostream& os) const {
    auto it = items();
    const std::uint32_t sz = static_cast<std::uint32_t>(it.size());
    os.write(reinterpret_cast<const char*>(&sz), sizeof(sz));
    for (auto [k, v] : it) {
        os.write(reinterpret_cast<const char*>(&k), sizeof(k));
        os.write(reinterpret_cast<const char*>(&v), sizeof(v));
    }
}

void HashTable::deserialize_binary(std::istream& is) {
    std::uint32_t sz = 0;
    is.read(reinterpret_cast<char*>(&sz), sizeof(sz));
    rehash(next_pow2(static_cast<int>(sz * 2 + 8)));
    for (std::uint32_t i = 0; i < sz; ++i) {
        int k = 0;
        int v = 0;
        is.read(reinterpret_cast<char*>(&k), sizeof(k));
        is.read(reinterpret_cast<char*>(&v), sizeof(v));
        insert(k, v);
    }
}

void HashTable::serialize_text(std::ostream& os) const {
    auto it = items();
    os << it.size() << "\n";
    for (auto [k, v] : it) {
        os << k << " " << v << "\n";
    }
}

void HashTable::deserialize_text(std::istream& is) {
    std::size_t sz = 0;
    if (!(is >> sz)) return;
    rehash(next_pow2(static_cast<int>(sz * 2 + 8)));
    for (std::size_t i = 0; i < sz; ++i) {
        int k = 0;
        int v = 0;
        is >> k >> v;
        insert(k, v);
    }
}


CuckooHashTable::CuckooHashTable(int capacity) {
    capacity_ = 1;
    while (capacity_ < capacity) capacity_ <<= 1;
    if (capacity_ < 2) capacity_ = 2;

    keys1_ = new int[capacity_];
    vals1_ = new int[capacity_];
    used1_ = new unsigned char[capacity_]();

    keys2_ = new int[capacity_];
    vals2_ = new int[capacity_];
    used2_ = new unsigned char[capacity_]();
}

CuckooHashTable::~CuckooHashTable() {
    delete[] keys1_;
    delete[] vals1_;
    delete[] used1_;
    delete[] keys2_;
    delete[] vals2_;
    delete[] used2_;
}

int CuckooHashTable::hash1(int key) const {
    std::uint32_t x = static_cast<std::uint32_t>(key);
    x *= 2654435761u;
    return static_cast<int>(x & static_cast<std::uint32_t>(capacity_ - 1));
}
int CuckooHashTable::hash2(int key) const {
    std::uint32_t x = static_cast<std::uint32_t>(key);
    x ^= x >> 16;
    x *= 2246822519u;
    x ^= x >> 13;
    return static_cast<int>(x & static_cast<std::uint32_t>(capacity_ - 1));
}

void CuckooHashTable::rehash(int new_capacity) {
    auto current = items();
    delete[] keys1_;
    delete[] vals1_;
    delete[] used1_;
    delete[] keys2_;
    delete[] vals2_;
    delete[] used2_;

    capacity_ = 1;
    while (capacity_ < new_capacity) capacity_ <<= 1;
    if (capacity_ < 2) capacity_ = 2;

    keys1_ = new int[capacity_];
    vals1_ = new int[capacity_];
    used1_ = new unsigned char[capacity_]();

    keys2_ = new int[capacity_];
    vals2_ = new int[capacity_];
    used2_ = new unsigned char[capacity_]();

    count_ = 0;
    for (auto [k, v] : current) {
        insert(k, v);
    }
}

bool CuckooHashTable::contains(int key) const { return get(key) != -1; }

int CuckooHashTable::get(int key) const {
    const int i1 = hash1(key);
    if (used1_[i1] && keys1_[i1] == key) return vals1_[i1];

    const int i2 = hash2(key);
    if (used2_[i2] && keys2_[i2] == key) return vals2_[i2];

    return -1;
}

void CuckooHashTable::insert(int key, int value) {
    if (contains(key)) {
        const int i1 = hash1(key);
        if (used1_[i1] && keys1_[i1] == key) {
            vals1_[i1] = value;
            return;
        }
        const int i2 = hash2(key);
        if (used2_[i2] && keys2_[i2] == key) {
            vals2_[i2] = value;
            return;
        }
    }

    if (static_cast<double>(count_ + 1) / static_cast<double>(capacity_) > 0.45) {
        rehash(capacity_ * 2);
    }

    int cur_k = key;
    int cur_v = value;

    constexpr int kMaxKicks = 32;
    for (int kick = 0; kick < kMaxKicks; ++kick) {
        int i1 = hash1(cur_k);
        if (!used1_[i1]) {
            used1_[i1] = 1;
            keys1_[i1] = cur_k;
            vals1_[i1] = cur_v;
            ++count_;
            return;
        }
        std::swap(cur_k, keys1_[i1]);
        std::swap(cur_v, vals1_[i1]);

        int i2 = hash2(cur_k);
        if (!used2_[i2]) {
            used2_[i2] = 1;
            keys2_[i2] = cur_k;
            vals2_[i2] = cur_v;
            ++count_;
            return;
        }
        std::swap(cur_k, keys2_[i2]);
        std::swap(cur_v, vals2_[i2]);
    }

    rehash(capacity_ * 2);
    insert(cur_k, cur_v);
}

void CuckooHashTable::remove(int key) {
    const int i1 = hash1(key);
    if (used1_[i1] && keys1_[i1] == key) {
        used1_[i1] = 0;
        --count_;
        return;
    }
    const int i2 = hash2(key);
    if (used2_[i2] && keys2_[i2] == key) {
        used2_[i2] = 0;
        --count_;
        return;
    }
}

std::vector<std::pair<int, int>> CuckooHashTable::items() const {
    std::vector<std::pair<int, int>> out;
    out.reserve(static_cast<std::size_t>(count_));
    for (int i = 0; i < capacity_; ++i) {
        if (used1_[i]) out.emplace_back(keys1_[i], vals1_[i]);
    }
    for (int i = 0; i < capacity_; ++i) {
        if (used2_[i]) out.emplace_back(keys2_[i], vals2_[i]);
    }
    return out;
}

void CuckooHashTable::serialize_binary(std::ostream& os) const {
    auto it = items();
    const std::uint32_t sz = static_cast<std::uint32_t>(it.size());
    os.write(reinterpret_cast<const char*>(&sz), sizeof(sz));
    for (auto [k, v] : it) {
        os.write(reinterpret_cast<const char*>(&k), sizeof(k));
        os.write(reinterpret_cast<const char*>(&v), sizeof(v));
    }
}

void CuckooHashTable::deserialize_binary(std::istream& is) {
    std::uint32_t sz = 0;
    is.read(reinterpret_cast<char*>(&sz), sizeof(sz));
    rehash(static_cast<int>(sz * 4 + 8));
    for (std::uint32_t i = 0; i < sz; ++i) {
        int k = 0;
        int v = 0;
        is.read(reinterpret_cast<char*>(&k), sizeof(k));
        is.read(reinterpret_cast<char*>(&v), sizeof(v));
        insert(k, v);
    }
}

void CuckooHashTable::serialize_text(std::ostream& os) const {
    auto it = items();
    os << it.size() << "\n";
    for (auto [k, v] : it) {
        os << k << " " << v << "\n";
    }
}

void CuckooHashTable::deserialize_text(std::istream& is) {
    std::size_t sz = 0;
    if (!(is >> sz)) return;
    rehash(static_cast<int>(sz * 4 + 8));
    for (std::size_t i = 0; i < sz; ++i) {
        int k = 0;
        int v = 0;
        is >> k >> v;
        insert(k, v);
    }
}
